File: packages/validation/src/manifest.ts
import { z } from 'zod';

/**
 * Versioned schema for an extension manifest (orbit.json). This is the trust
 * boundary between third-party extension authors and the host: every field is
 * validated, lengths are bounded, and identifiers are constrained to safe
 * character sets so a malicious manifest cannot smuggle traversal sequences,
 * shell metacharacters or absurd values into the host.
 */

export const MANIFEST_SCHEMA_VERSION = 1;

/** Safe identifier: lowercase, alnum, dashes; no dots/slashes/spaces. */
const identifier = z
  .string()
  .min(1)
  .max(64)
  .regex(/^[a-z0-9][a-z0-9-]*$/, 'must be lowercase alphanumeric with dashes');

const semver = z
  .string()
  .regex(/^\d+\.\d+\.\d+(?:-[0-9A-Za-z-.]+)?(?:\+[0-9A-Za-z-.]+)?$/, 'must be semver');

const permissionId = z.enum([
  'clipboard.read',
  'clipboard.write',
  'files.read',
  'files.write',
  'apps.launch',
  'apps.enumerate',
  'window.manage',
  'system.commands',
  'network',
  'selected-text.read',
  'text.insert',
  'browser.context',
  'calendar.read',
  'calendar.write',
  'ai',
  'shell.execute',
  'secure-storage',
]);

const commandMode = z.enum([
  'no-view',
  'list',
  'grid',
  'detail',
  'form',
  'menu-bar',
  'background',
  'interval',
  'ai-tool',
]);

const preferenceSchema = z.object({
  name: identifier,
  title: z.string().min(1).max(120),
  description: z.string().max(500).optional(),
  type: z.enum(['textfield', 'password', 'checkbox', 'dropdown', 'file', 'directory']),
  required: z.boolean().default(false),
  default: z.union([z.string(), z.boolean(), z.number()]).optional(),
  // Dropdown options
  data: z
    .array(z.object({ title: z.string().max(120), value: z.string().max(200) }))
    .max(100)
    .optional(),
});

const commandSchema = z.object({
  name: identifier,
  title: z.string().min(1).max(120),
  description: z.string().max(500).optional(),
  mode: commandMode,
  // Per-command keywords
  keywords: z.array(z.string().max(40)).max(50).optional(),
  // Interval (seconds) for interval commands
  interval: z.number().int().min(10).max(86_400).optional(),
  preferences: z.array(preferenceSchema).max(50).optional(),
});

export const manifestSchema = z.object({
  $schema: z.literal(MANIFEST_SCHEMA_VERSION).or(z.undefined()),
  name: identifier,
  title: z.string().min(1).max(120),
  description: z.string().min(1).max(1000),
  author: identifier,
  version: semver,
  icon: z
    .string()
    .max(200)
    // icon is a packaged relative path; reject traversal & absolute paths
    .refine((p) => !p.includes('..') && !p.startsWith('/') && !/^[a-zA-Z]:/.test(p), {
      message: 'icon must be a relative path without traversal',
    })
    .optional(),
  categories: z.array(z.string().max(40)).max(10).default([]),
  platforms: z.array(z.enum(['windows', 'macos', 'linux'])).min(1).max(3),
  commands: z.array(commandSchema).min(1).max(100),
  preferences: z.array(preferenceSchema).max(50).optional(),
  permissions: z.array(permissionId).max(20).default([]),
  dependencies: z.record(z.string().max(100), z.string().max(100)).optional(),
  minOrbitVersion: semver.optional(),
  website: z.string().url().max(300).optional(),
  repository: z.string().url().max(300).optional(),
  license: z.string().max(60).optional(),
  changelog: z.string().max(20_000).optional(),
});

export type Manifest = z.infer<typeof manifestSchema>;
export type ManifestCommand = z.infer<typeof commandSchema>;
export type ManifestPreference = z.infer<typeof preferenceSchema>;

export interface ValidationOk {
  readonly ok: true;
  readonly manifest: Manifest;
}
export interface ValidationErr {
  readonly ok: false;
  readonly errors: ReadonlyArray<{ path: string; message: string }>;
}

/** Validate an unknown value as a manifest, returning a typed result. */
export function validateManifest(input: unknown): ValidationOk | ValidationErr {
  const parsed = manifestSchema.safeParse(input);
  if (parsed.success) return { ok: true, manifest: parsed.data };
  return {
    ok: false,
    errors: parsed.error.issues.map((i) => ({
      path: i.path.join('.'),
      message: i.message,
    })),
  };
}

/** Cross-check that every command's declared permissions exist in the manifest. */
export function permissionsAreDeclared(manifest: Manifest): boolean {
  // (Reserved for future per-command permission narrowing; today permissions are
  // declared at the manifest level and apply to all commands.)
  return manifest.permissions.length <= 20;
}

File: crates/orbit-extensions/src/manifest.rs
//! Extension manifest (`manifest.json`) parsing + validation.
//!
//! This is the host-side trust boundary for third-party extensions: every field
//! is bounded and identifiers are constrained so a malicious manifest can't
//! smuggle traversal, odd command modes, or unknown permissions past the host.
//! Mirrors the renderer-side `@orbit/validation` manifest schema.

use serde::Deserialize;

/// Command modes the host understands (a subset of the renderer's enum that the
/// runtime can actually execute today).
pub const ALLOWED_MODES: &[&str] = &["no-view", "list"];

/// Permissions the broker can enforce.
pub const ALLOWED_PERMISSIONS: &[&str] = &[
    "clipboard.read",
    "clipboard.write",
    "files.read",
    "files.write",
    "apps.launch",
    "apps.enumerate",
    "window.manage",
    "system.commands",
    "network",
    "selected-text.read",
    "text.insert",
    "browser.context",
    "calendar.read",
    "calendar.write",
    "ai",
    "shell.execute",
    "secure-storage",
];

#[derive(Debug, Clone, Deserialize, PartialEq, Eq)]
pub struct CommandDef {
    pub name: String,
    pub title: String,
    #[serde(default)]
    pub description: Option<String>,
    pub mode: String,
    #[serde(default)]
    pub keywords: Vec<String>,
}

fn default_main() -> String {
    "index.mjs".to_string()
}

#[derive(Debug, Clone, Deserialize, PartialEq, Eq)]
pub struct Manifest {
    pub name: String,
    pub title: String,
    #[serde(default)]
    pub description: String,
    #[serde(default)]
    pub version: String,
    #[serde(default)]
    pub author: String,
    #[serde(default)]
    pub permissions: Vec<String>,
    pub commands: Vec<CommandDef>,
    /// Entry script (relative, no traversal). Defaults to `index.mjs`.
    #[serde(default = "default_main")]
    pub main: String,
}

fn is_safe_identifier(s: &str) -> bool {
    !s.is_empty()
        && s.len() <= 64
        && s.bytes().next().is_some_and(|b| b.is_ascii_lowercase() || b.is_ascii_digit())
        && s.bytes().all(|b| b.is_ascii_lowercase() || b.is_ascii_digit() || b == b'-')
}

fn is_safe_main(s: &str) -> bool {
    !s.is_empty()
        && s.len() <= 200
        && !s.contains("..")
        && !s.starts_with('/')
        && !s.starts_with('\\')
        && !s.contains(':') // rejects C:\ and url-like
}

impl Manifest {
    /// Validate all bounds/identifiers. Returns a human-readable error on failure.
    pub fn validate(&self) -> Result<(), String> {
        if !is_safe_identifier(&self.name) {
            return Err("manifest.name must be lowercase alphanumeric with dashes (<=64)".into());
        }
        if self.title.trim().is_empty() || self.title.chars().count() > 120 {
            return Err("manifest.title must be 1–120 chars".into());
        }
        if self.commands.is_empty() {
            return Err("manifest must declare at least one command".into());
        }
        if self.commands.len() > 100 {
            return Err("too many commands".into());
        }
        for c in &self.commands {
            if !is_safe_identifier(&c.name) {
                return Err(format!("command name '{}' is not a safe identifier", c.name));
            }
            if c.title.trim().is_empty() || c.title.chars().count() > 120 {
                return Err(format!("command '{}' title must be 1–120 chars", c.name));
            }
            if !ALLOWED_MODES.contains(&c.mode.as_str()) {
                return Err(format!(
                    "command '{}' mode '{}' is not supported (use {:?})",
                    c.name, c.mode, ALLOWED_MODES
                ));
            }
        }
        if self.permissions.len() > 20 {
            return Err("too many permissions".into());
        }
        for p in &self.permissions {
            if !ALLOWED_PERMISSIONS.contains(&p.as_str()) {
                return Err(format!("unknown permission '{p}'"));
            }
        }
        if !is_safe_main(&self.main) {
            return Err("manifest.main must be a relative path without traversal".into());
        }
        Ok(())
    }
}

/// Parse + validate a manifest from JSON text.
pub fn parse(json: &str) -> Result<Manifest, String> {
    let manifest: Manifest = serde_json::from_str(json).map_err(|e| format!("invalid manifest JSON: {e}"))?;
    manifest.validate()?;
    Ok(manifest)
}

#[cfg(test)]
mod tests {
    use super::*;

    const VALID: &str = r#"{
        "name":"dev-utils","title":"Dev Utils","description":"d","version":"0.1.0",
        "author":"orbit","permissions":["clipboard.write"],
        "commands":[{"name":"gen","title":"Generate","mode":"no-view"},
                    {"name":"hist","title":"History","mode":"list"}]
    }"#;

    #[test]
    fn parses_valid_manifest() {
        let m = parse(VALID).unwrap();
        assert_eq!(m.name, "dev-utils");
        assert_eq!(m.commands.len(), 2);
        assert_eq!(m.main, "index.mjs"); // default
        assert!(m.permissions.contains(&"clipboard.write".to_string()));
    }

    #[test]
    fn rejects_bad_identifier() {
        let bad = VALID.replace("\"name\":\"dev-utils\"", "\"name\":\"Dev Utils!\"");
        assert!(parse(&bad).is_err());
    }

    #[test]
    fn rejects_unknown_permission() {
        let bad = VALID.replace("clipboard.write", "root.everything");
        assert!(parse(&bad).is_err());
    }

    #[test]
    fn rejects_unsupported_mode() {
        let bad = VALID.replace("\"mode\":\"list\"", "\"mode\":\"menu-bar\"");
        assert!(parse(&bad).is_err());
    }

    #[test]
    fn rejects_traversal_in_main() {
        let bad = VALID.replace(
            "\"commands\"",
            "\"main\":\"../../evil.mjs\",\"commands\"",
        );
        assert!(parse(&bad).is_err());
    }

    #[test]
    fn rejects_empty_commands() {
        let bad = r#"{"name":"x","title":"X","commands":[]}"#;
        assert!(parse(bad).is_err());
    }
}
